src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs

```diff
@@ -145,25 +136,35 @@ internal IServiceCallSite GetResolveCallSite(IService service, ISet<Type> callSi
 
         public void Dispose()
         {
-            var disposables = Interlocked.Exchange(ref _disposables, null);
-
-            if (disposables != null)
+            lock (_disposables)
             {
-                foreach (var disposable in disposables)
+                foreach (var disposable in _disposables)
                 {
                     disposable.Dispose();
                 }
+
+                _disposables.Clear();
+                _resolvedServices.Clear();
             }
         }
 
         private object CaptureDisposable(object service)
         {
+            // Skip capturing disposables for the root
+            if (_root == this)
+            {
+                return service;
+            }
+
             if (!object.ReferenceEquals(this, service))
             {
                 var disposable = service as IDisposable;
                 if (disposable != null)
                 {
-                    _disposables.Add(disposable);
+                    lock (_disposables)
```

Comment:
Not really, but we have no tests to verify it either. We don't even seem to check if we call GetService from a disposed service provider either.